src/uniformscalingmap.jl

```diff
@@ -19,20 +19,48 @@ LinearAlgebra.transpose(A::UniformScalingMap) = A
 LinearAlgebra.adjoint(A::UniformScalingMap)   = UniformScalingMap(conj(A.λ), size(A))
 
 # multiplication with vector
+Base.:(*)(A::UniformScalingMap, x::AbstractVector) = A.λ * x
 # call of LinearAlgebra.generic_mul! since order of arguments in mul! in stdlib/LinearAlgebra/src/generic.jl
 # TODO: either leave it as is or use mul! (and lower bound on version) once fixed in LinearAlgebra
-A_mul_B!(y::AbstractVector, A::UniformScalingMap, x::AbstractVector) =
-    (length(x) == length(y) == A.M ? LinearAlgebra.generic_mul!(y, A.λ, x) : throw(DimensionMismatch("A_mul_B!")))
-Base.:(*)(A::UniformScalingMap, x::AbstractVector) = A.λ * x
+function A_mul_B!(y::AbstractVector, A::UniformScalingMap, x::AbstractVector)
+    (length(x) == length(y) == A.M || throw(DimensionMismatch("A_mul_B!")))
+    if iszero(A.λ)
+        return fill!(y, 0)
+    elseif isone(A.λ)
+        return copyto!(y, x)
+    else
+        return LinearAlgebra.generic_mul!(y, A.λ, x)
+    end
+end
 
-At_mul_B!(y::AbstractVector, A::UniformScalingMap, x::AbstractVector) =
-    (length(x) == length(y) == A.M ? LinearAlgebra.generic_mul!(y, A.λ, x) : throw(DimensionMismatch("At_mul_B!")))
+function At_mul_B!(y::AbstractVector, A::UniformScalingMap, x::AbstractVector)
+    (length(x) == length(y) == A.M || throw(DimensionMismatch("At_mul_B!")))
+    if iszero(A.λ)
+        return fill!(y, 0)
+    elseif isone(A.λ)
+        return copyto!(y, x)
+    else
+        return LinearAlgebra.generic_mul!(y, A.λ, x)
+    end
+end
 
-Ac_mul_B!(y::AbstractVector, A::UniformScalingMap, x::AbstractVector) =
-    (length(x) == length(y) == A.M ? LinearAlgebra.generic_mul!(y, conj(A.λ), x) : throw(DimensionMismatch("Ac_mul_B!")))
+function Ac_mul_B!(y::AbstractVector, A::UniformScalingMap, x::AbstractVector)
+    (length(x) == length(y) == A.M || throw(DimensionMismatch("At_mul_B!")))
+    if iszero(A.λ)
+        return fill!(y, 0)
+    elseif isone(A.λ)
+        return copyto!(y, x)
+    else
+        return LinearAlgebra.generic_mul!(y, conj(A.λ), x)
+    end
+end
 
 # combine LinearMap and UniformScaling objects in linear combinations
-Base.:(+)(A1::LinearMap, A2::UniformScaling{T}) where {T} = A1 + UniformScalingMap{T}(A2[1,1], size(A1, 1))
-Base.:(+)(A1::UniformScaling{T}, A2::LinearMap) where {T} = UniformScalingMap{T}(A1[1,1], size(A2, 1)) + A2
-Base.:(-)(A1::LinearMap, A2::UniformScaling{T}) where {T} = A1 - UniformScalingMap{T}(A2[1,1], size(A1, 1))
-Base.:(-)(A1::UniformScaling{T}, A2::LinearMap) where {T} = UniformScalingMap{T}(A1[1,1], size(A2, 1)) - A2
+Base.:(+)(A1::LinearMap{T}, A2::UniformScaling) where {T} =
+    A1 + UniformScalingMap(convert(T, A2[1,1]), size(A1, 1))
```

Comment:
This looks wrong; there is no guarantee that `A2[1,1]` (i.e. lambda) can be converted to type `T`. Some promotion should be involved at least.

Reply:
I'm not sure I fully understand the difference between conversion and promotion. The first sentence in the Julia docs for promotion reads

> Promotion refers to converting values of mixed types to a single common type.

To me, this sounds like first the types of the arguments of `promote` are extracted, then the types are promoted, and eventually all arguments are converted to the common type. Do you have an example of what would not work with `convert`, but with `promote`?

Reply:
yes, if `T<:Real` and the `UniformScaling` parameter is complex, you will get an error.

If `T == Float32` and the `UniformScaling` is `Float64`, promotion would result in both being `FLoat64`, whereas the current approach will convert to `Float32`.

Reply:
Ok, thanks, now I see. So we do something like
```julia
convert(promote_type(T, eltype(A2)), A2[1,1])
```
? And maybe I'll replace the `A2[1,1]` by `A2.λ`.

Reply:
Sure, but the old version was also working fine, or not? I am fine either way, will amount to more or less the same thing.

Reply:
Yes it was, but the lambda is more direct... but probably this would get compiled away anyway. 😄